Boolean_set_operations_2/include/CGAL/Boolean_set_operations_2/Bso_internal_functions.h

```diff
@@ -8,192 +8,522 @@
+  // Use the first polygon to determine the (default) traits
+  typedef typename Gps_polyline_traits<Pgn1>::Polyline_traits   Polyline_traits;
+  typedef General_polygon_2<Polyline_traits>                    General_pgn;
+  typedef General_polygon_with_holes_2<General_pgn>             General_pwh;
+
+  General_pwh general_pwh;
+  typename Gps_polyline_traits<Pgn1>::Traits traits;
+  const Polyline_traits& ptraits(traits);
+  auto res = s_join(convert_polygon(pgn1, ptraits),
+                    convert_polygon(pgn2, ptraits),
+                    general_pwh, traits);
+  pwh = convert_polygon_back<Kernel, Container>(general_pwh);
+  return res;
+}
+
+// Range
+// With traits
+template <typename InputIterator, typename OutputIterator, typename Traits>
+inline OutputIterator r_join(InputIterator begin, InputIterator end,
+                             OutputIterator oi, Traits& traits,
+                             unsigned int k=5) {
+  if (begin == end) return oi;
+  General_polygon_set_2<Traits> gps(*begin, traits);
+  gps.join(std::next(begin), end, k);
+  return gps.polygons_with_holes(oi);
+}
+
+// Without traits
+template <typename InputIterator, typename OutputIterator>
+inline OutputIterator r_join(InputIterator begin, InputIterator end,
+                             OutputIterator oi, unsigned int k=5) {
+  typedef typename std::iterator_traits<InputIterator>::value_type Pgn;
+  typename Gps_polyline_traits<Pgn>::Traits traits;
+  const typename Gps_polyline_traits<Pgn>::Polyline_traits& ptraits(traits);
+  return r_join(convert_polygon_iterator(begin, ptraits),
+                convert_polygon_iterator(end, ptraits),
+                convert_polygon_back(oi, *begin), traits, k);
+}
+
+// Double Range
+// With traits
+template <typename InputIterator1, typename InputIterator2,
+          typename OutputIterator, typename Traits>
+inline OutputIterator r_join(InputIterator1 begin1, InputIterator1 end1,
+                             InputIterator2 begin2, InputIterator2 end2,
+                             OutputIterator oi, Traits& traits,
+                             unsigned int k=5) {
+  if (begin1 == end1) return r_join(begin2, end2, oi, traits, k);
+  General_polygon_set_2<Traits> gps(*begin1, traits);
+  gps.join(std::next(begin1), end1, begin2, end2, k);
+  return gps.polygons_with_holes(oi);
+}
+
+// Without traits
+template <typename InputIterator1, typename InputIterator2,
+          typename OutputIterator>
+inline OutputIterator r_join(InputIterator1 begin1, InputIterator1 end1,
+                             InputIterator2 begin2, InputIterator2 end2,
+                             OutputIterator oi, unsigned int k=5) {
+  typedef typename std::iterator_traits<InputIterator1>::value_type Pgn;
+  typename Gps_polyline_traits<Pgn>::Traits traits;
+  const typename Gps_polyline_traits<Pgn>::Polyline_traits& ptraits(traits);
+  return r_join(convert_polygon_iterator(begin1, ptraits),
```

Comment:
Here as well we have to do something special in case of  `begin == end`, but I do not yet what.